# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [2]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [ ]:
# YOUR CODE HERE
URL = "https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population"
HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}

response = requests.get(URL, headers=HEADERS)

Use Beautiful Soup to parse this string into a tree called `soup`

In [ ]:
# YOUR CODE HERE
soup = BeautifulSoup(response.text, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [ ]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [ ]:
len(soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      "style": "text-align:right"}
                  ))

1

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [ ]:
table = soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      "style": "text-align:right"}
                  )[0]


**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
cities = table.find_all("tr")
len(cities)

348

In [ ]:
city = table.find_all("tr")[2]
city

<tr>
<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>
<td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
</td>
<td style="text-align:right;">8,478,072
</td>
<td style="text-align:right;">8,804,190
</td>
<td style="text-align:right;"><span data-sort-value="2999630000000000000♠" style="display:none"></span><span style="color:red">−3.70%</span>
</td>
<td style="text-align:right;">300.5
</td>
<td style="text-align:right;">778.3
</td>
<td style="text-align:right;">29,298
</td>
<td style="text-align:right;">11,312
</td>
<td><small><span class="geo-inline"><style data-mw-deduplicate="TemplateStyles:r1156832818">.mw-parser-output .geo-default,.mw-parser-output .geo-dms,.mw-parser-output .geo-dec{display:inline}.mw-parser-output .geo-nondefault,.mw-parser-output .geo-mult

In [ ]:
cells = table.find_all("td")
cells

[<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
 </td>,
 <td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
 </td>,
 <td style="text-align:right;">8,478,072
 </td>,
 <td style="text-align:right;">8,804,190
 </td>,
 <td style="text-align:right;"><span data-sort-value="2999630000000000000♠" style="display:none"></span><span style="color:red">−3.70%</span>
 </td>,
 <td style="text-align:right;">300.5
 </td>,
 <td style="text-align:right;">778.3
 </td>,
 <td style="text-align:right;">29,298
 </td>,
 <td style="text-align:right;">11,312
 </td>,
 <td><small><span class="geo-inline"><style data-mw-deduplicate="TemplateStyles:r1156832818">.mw-parser-output .geo-default,.mw-parser-output .geo-dms,.mw-parser-output .geo-dec{display:inline}.mw-parser-output .geo-nondefault,.mw-

In [ ]:
#city
name = cells[0].get_text(strip = True)
name

'New York[c]'

In [ ]:
# state
state = cells[1].get_text(strip = True)
state

'NY'

In [ ]:
population = cells[2].get_text(strip = True)
population

'8,478,072'

In [ ]:
land_area = cells[5].get_text(strip = True)
land_area

'300.5'

In [ ]:
# city, state, population, land size
# initialize an empty list
rows = []

# iterate over all rows in the faculty table
for city in table.find_all("tr")[2:]:

    # Get all the cells (<td>) in the row.
    cells = city.find_all("td")

    # The information we need is the text between tags.


    city = cells[0].get_text(strip = True)
    city

    state = cells[1].get_text(strip = True)
    state

    population = cells[2].get_text(strip = True)
    population

    area = cells[5].get_text(strip = True)
    area

    rows.append({
        "city": city,
        "state": state,
        "population": population,
        "area": area
    })

In [ ]:
pd.DataFrame(rows)

city state population   area
0    New York[c]    NY  8,478,072  300.5
1    Los Angeles    CA  3,878,704  469.5
2        Chicago    IL  2,721,308  227.7
3        Houston    TX  2,390,125  640.4
4        Phoenix    AZ  1,673,164  518.0
..           ...   ...        ...    ...
341      Deltona    FL    100,513   37.3
342  Federal Way    WA    100,252   22.3
343   San Angelo    TX    100,159   59.7
344        Tracy    CA    100,136   25.9
345      Sunrise    FL    100,128   16.2

[346 rows x 4 columns]

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas - section skipped



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [1]:
# pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [2]:
# pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable sortable', "style": "text-align:center"})

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [3]:
# df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable sortable', "style": "text-align:center"})[0]
# df_cities2

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [4]:
courses = pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")
courses

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
df = soup.find_all("div", {"class": "courseblock"})
len(df)

10


In [7]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# Step 1: Load the webpage
url = "http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

# Step 2: Find all course blocks
course_blocks = soup.find_all("div", class_="courseblock")

# Step 3: Prepare lists
course_names = []
terms_offered = []

# Step 4: Loop through all courses
for block in course_blocks:
    # Find course name/number
    title_tag = block.find("p", class_="courseblocktitle")
    if title_tag:
        # Remove the <span> with units (e.g., “(4 units)”) if present
        span = title_tag.find("span")
        if span:
            span.extract()
        course_name = title_tag.get_text(strip=True)

        # Only keep DATA or STAT courses
        if course_name.startswith(("DATA", "STAT")):
            # Find term offered info
            term_tag = block.find("p", class_="noindent")
            term = term_tag.get_text(strip=True) if term_tag else None

            course_names.append(course_name)
            terms_offered.append(term)

# Step 5: Build DataFrame
df = pd.DataFrame({
    "Course": course_names,
    "Term Offered": terms_offered
})

# Step 6: Check results
print(f"Total rows: {len(df)}")
df.head()

Total rows: 74


Course  \
0                  DATA 100. Data Science for All I.   
1            DATA 301. Introduction to Data Science.   
2         DATA 401. Data Science Process and Ethics.   
3  DATA 402. Mathematical Foundations of Data Sci...   
4        DATA 403. Data Science Projects Laboratory.   

                       Term Offered  
0  Term Typically Offered: F, W, SP  
1  Term Typically Offered: F, W, SP  
2         Term Typically Offered: F  
3         Term Typically Offered: F  
4         Term Typically Offered: F

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag